libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationApp.kt

```diff
@@ -77,6 +77,25 @@ object MapboxNavigationApp {
     @JvmStatic
     fun isSetup(): Boolean = mapboxNavigationAppDelegate.isSetup
 
+    /**
+     * When calling [setup] multiple times, all registered [MapboxNavigationObserver] will be
+     * detached from the instance of [MapboxNavigation] that is being destroyed. This is needed for
+     * maintaining state across options changes. You do not need to clear the state when
+     * [isOptionsChanging] is true.
+     */
+    @UiThread
+    @JvmStatic
+    fun isOptionsChanging(): Boolean = mapboxNavigationAppDelegate.isOptionsChanging()
+
+    /**
+     * When observing [MapboxNavigation] from a [Lifecycle], the [Lifecycle.Event] may be triggered
+     * because of a configuration change. This will help you know if the lifecycle events are
+     * triggered events like mobile device orientation changes.
+     */
+    @UiThread
+    @JvmStatic
+    fun isConfigurationChanging(): Boolean = mapboxNavigationAppDelegate.isConfigurationChanging()
```

Comment:
How do you envision using these methods? Do you have an example?
I can't say about options changing but regarding configuration changing I'm not sure if that's the way to go.
I'm worried that we might use this method somewhere and when it returns true we'll just ignore some actions, meanwhile the state will become inconsistent because some other components will be recreated.
I think it would be more neat and clear if the components that don't care about configuration changes just had a broader lifecycle. It's basically your first idea from https://github.com/mapbox/mapbox-navigation-android/pull/6766#issuecomment-1363574672.

Reply:
Yeah I have the same concern. I've been debating whether the MapboxNavigationObserver implementation should ignore the detach when configuration changes. Or if MapboxNavigaitonApp should not detach observers when configuration is changing.

I think we need to support the case where, a MapboxNavigationObserver is only attached while in landscape mode. So I've been thinking it should be the MapboxNavigationObserver responsibility to create a definition and usage requirements

Reply:
There are issues with putting the logic in MapboxNavigationObserver. For example, if we want the observer to be usable without MapboxNavigationApp.

Reply:
So another option am considering, is what if you can specify what you want during registration. For example, add a new way to register an observer

MapboxNavigationApp.registerStrongObserver(observer)

the strong (there is probably a better word) observer would not detach during configuration or option changes. another option is to add parameters to the registerObserver function.

Reply:
I actually like the strong observer idea. And I think that a new type of observer is more explicit than passing the options as an argument. It's also less flexible but I'm not sure if we're gonna need a lot of flexibility here.
And I guess we should also support the case when we want to do A on `onAttached`, B on configuration/options change and C on `onDetached`. We'd probably need two different observers to resolve this. To make it possible we should guarantee the order in which the observers are called. For example, if we register one `MapboxNavigationObserver` (O1) and `MapboxNavigationStrongObserver` (O2), then common `onAttached`/`onDetached` should be passed first to O1, then to O2 (or whatever order is more convenient). WDYT?

Reply:
> I actually like the strong observer idea.

👍 

> And I think that a new type of observer is more explicit than passing the options as an argument

Do you mean, preferring 2 over 1 like this? I can see it, but there are pros and cons 🤔 
1. MapboxNavigationApp.registerStrongObserver(object : MapboxNavigationObserver ..)
2. MapboxNavigationApp.registerObserver(object : MapboxNavigationStrongObserver ..)
3. Add MapboxNavigationObserver#isStrong() overridable configuration
4. Add MapboxNavigationObserver#onConfigurationChanged() and then make all observers strong
5. Add MapboxNavigationApp#isConfigurationChanging() so that onAttached/onDetached can figure it out

I'm not sure what I prefer yet.. it seems all of them technically work so we could list some pros/cons to decide.

> And I guess we should also support the case when we want to do A on onAttached, B on configuration/options change and C on onDetached.

We can also consider, adding an overridable function to MapboxNavigationObserver (e.g., onConfigurationChanged..).  The pull request as it is now, requires a condition to see if onAttached/onDetached is triggered by a configuration change.

> MapboxNavigationObserver (O1) and MapboxNavigationStrongObserver (O2), then common onAttached/onDetached should be passed first to O1, then to O2 (or whatever order is more convenient). WDYT?

I'm not yet sold on `MapboxNavigationStrongObserver` because I bet if we listed the pros and cons, we may prefer solution 3 to add an overridable configuration to `MapboxNavigationObserver`, or we may decide on something else. I don't like that we would have to change the type of the observer. I would like to at least start with the flexibility to be able to make any `MapboxNavigationObserver` a strong observer.

Reply:
> I don't like that we would have to change the type of the observer.

Why? We are intoducing a new type of observer with behaviour that was not present before. We could do with a new interface here.

> Add MapboxNavigationObserver#isStrong() overridable configuration

How are you going to implement it without breaking the API? There were some problems with introducing an interface method with default implementation. Converting it to an abstract class will break java users for sure.

Reply:
> > I don't like that we would have to change the type of the observer.
> Why? We are intoducing a new type of observer with behaviour that was not present before. We could do with a new interface here.

There is no difference with the implementation of the interface. Maybe we do this approach, just think we should consider other approaches first.

We can change the `MapboxNavigationObserver` implementation if we convert the interface to a java interface ([an example here](https://github.com/mapbox/mapbox-navigation-android/pull/6420)). Default functions are supported in java, default functions will have better support in kotlin after version `1.6.20` https://youtrack.jetbrains.com/issue/KT-54239